provd/internal/services/gdm/gdm.go

```diff
@@ -0,0 +1,148 @@
+	}
+
+	// Ignoring the error here, since GDM's private bus doesn't support the standard "Hello" method. It is needed for the tests, though.
+	_ = s.privateConn.Hello()
+
+	// Ping the UserVerifier and Greeter objects
+	s.userVerifier = s.privateConn.Object(consts.DbusGdmPrefix+".UserVerifier", dbus.ObjectPath("/org/gnome/DisplayManager/Session"))
+	err = s.userVerifier.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping UserVerifier: %s", err)
+	}
+	s.greeter = s.privateConn.Object(consts.DbusGdmPrefix+".Greeter", dbus.ObjectPath("/org/gnome/DisplayManager/Session"))
+	err = s.greeter.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping Greeter: %s", err)
+	}
+
+	return s, nil
+}
+
+// LaunchDesktopSession launches a desktop session for the given user and password.
+func (s *Service) LaunchDesktopSession(ctx context.Context, req *pb.GdmServiceRequest) (*emptypb.Empty, error) {
+	// Validate request
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "received a nil request")
+	}
+	if req.Username == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "received an empty username")
+	}
+
+	// Listen for signals on the session bus
+	signals := make(chan *dbus.Signal)
+	defer close(signals)
+	s.privateConn.Signal(signals)
+	defer s.privateConn.RemoveSignal(signals)
+
+	// Begin user verification
+	s.userVerifier.Call(consts.DbusGdmPrefix+".UserVerifier.BeginVerificationForUser", 0, "gdm-password", req.Username)
+
+	// Wait for signals
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, status.Errorf(codes.DeadlineExceeded, "context deadline exceeded")
+		case signal := <-signals:
+			switch signal.Name {
+			// Problem signal
+			case consts.DbusGdmPrefix + ".UserVerifier.Problem":
+				slog.Error(fmt.Sprintf("Problem signal received: %#v", signal))
+				return nil, status.Errorf(codes.Internal, "received Problem signal: %v", signal.Body[1])
+			// SecretInfoQuery signal
+			case consts.DbusGdmPrefix + ".UserVerifier.SecretInfoQuery":
+				slog.Debug(fmt.Sprintf("SecretInfoQuery signal received: %#v", signal))
+				msg, ok := signal.Body[0].(string)
+				if !ok {
+					return nil, status.Errorf(codes.Internal, "failed to get message from SecretInfoQuery signal")
+				}
+				// Reply with password
+				s.userVerifier.Call(consts.DbusGdmPrefix+".UserVerifier.AnswerQuery", 0, msg, req.Password)
+			// SessionOpened signal
+			case consts.DbusGdmPrefix + ".Greeter.SessionOpened":
+				slog.Debug(fmt.Sprintf("SessionOpened signal received: %#v", signal))
+				msg, ok := signal.Body[0].(string)
+				if !ok {
+					return nil, status.Errorf(codes.Internal, "failed to get message from SessionOpened signal")
+				}
+				// Tell GDM to start the session and return
+				s.greeter.Call(consts.DbusGdmPrefix+".Greeter.StartSessionWhenReady", 0, msg, true)
```

Comment:
1: What you could do with your calls to `StastSessionWhenReady` and `AnswerQuery` is look for returned from dbus and catch the cases where your side effects don`t succeed.

Reply:
Thanks, very good point - will do!